### PR TITLE
Allow context to work without controller

### PR DESCRIPTION
**Pull request by stephenprater**
The context constructor doesn't REQUIRE you to pass in a controller, but it breaks if you don't.  Since there's actually code to support it not existing, if we check to make sure we have it before we call stuff on it, we should be fine.

### CODE REVIEW

File: lib/blackbeard/context.rb [hunk 1]
```diff
@@ -61,14 +61,19 @@ def user_id
     end
 
     def visitor_id
-      @visitor_id ||= (controller.request.cookies['bbd'] || generate_visitor_id).to_i
+      @visitor_id ||=
+        if controller
+          controller.request.cookies['bbd'].to_i
```
Comment (rgraff) [inline lib/blackbeard/context.rb]: is this functionality change intentional? If the cookie is empty the visitor will be 0 which doesn't seem right.

Comment (stephenprater) [inline lib/blackbeard/context.rb]: You are correct and it has now been fixed.

**Comment by stephenprater** [comment]
@rgraff the problem is that without a request (or something that acts like a request) you can't obtain a Context.  So if you're in a worker or something where you DO have a user ID, but not a request, you won't be able to get a Context for any user unless you pass a faked-up request object.

**Comment by rgraff** [comment]
And now the bug you're trying to fix is painfully obvious. +2

**Comment by stephenprater** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;OqhHta5N&#x2F;901-improve-performance-of-new-collab-to-acceptable-levels"> Improve performance of New Collab to acceptable Levels</a>

**Pull request merged**